plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/session-offset-high-water-mark.ts

```diff
@@ -0,0 +1,95 @@
+import { captureException } from '@sentry/node'
+import { Redis } from 'ioredis'
+
+import { RedisPool } from '../../../../types'
+import { timeoutGuard } from '../../../../utils/db/utils'
+import { status } from '../../../../utils/status'
+
+const offsetHighWaterMarkKey = (sessionId: string) => `@posthog/replay/partition-high-water-marks/${sessionId}`
+
+/**
+ * If a file is written to S3 we need to know the offset of the last message in that file so that we can
+ * commit it to Kafka. But we don't write every offset as we bundle files to reduce the number of writes.
+ *
+ * And not every attempted commit will succeed
+ *
+ * That means if a consumer restarts or a rebalance moves a partition to another consumer we need to know
+ * which offsets have been written to S3 for each session, and which haven't
+ * so that we don't re-process those messages.
+ */
+export class SessionOffsetHighWaterMark {
+    constructor(private redisPool: RedisPool) {}
+
+    private async run<T>(description: string, fn: (client: Redis) => Promise<T>): Promise<T | null> {
+        const client = await this.redisPool.acquire()
+        const timeout = timeoutGuard(`${description} delayed. Waiting over 30 seconds.`)
+        try {
+            return await fn(client)
+        } catch (error) {
+            if (error instanceof SyntaxError) {
+                // invalid JSON
+                return null
+            } else {
+                throw error
+            }
+        } finally {
+            clearTimeout(timeout)
+            await this.redisPool.release(client)
+        }
+    }
+
+    public async set(sessionId: string, offset: number): Promise<void> {
+        const key = offsetHighWaterMarkKey(sessionId)
+        try {
+            await this.run(`write offset high-water mark ${key} `, async (client) => {
+                const pipeline = client.pipeline()
+                pipeline.set(key, offset)
+                // expire after 7 days
+                // if we can't process a session in 7 days we have bigger problems
+                // than re-processing some files
```

Comment:
How many records will this make in the worst case 🤔 millions for sure. Probably fine?